### PR TITLE
Add pyright language server for Python

### PR DESCRIPTION
**Pull request by zeronone**
Add support for Pyright language server (Python) as discussed #1863

### CODE REVIEW

**Comment by leungbk** [comment]
@andrew-christianson created a PR that he asked to revert over at the lsp-python-ms repo: https://github.com/emacs-lsp/lsp-python-ms/pull/110

Maybe you guys could collaborate on this. He had been planning to offer the option to disable definition-jumping with `pyright` and use the C# server for that instead since `pyright`'s definition-jumping behavior is a bit different.

**Comment by zeronone** [comment]
> @andrew-christianson created a PR that he asked to revert over at the lsp-python-ms repo: emacs-lsp/lsp-python-ms#110

Oh sorry, I didn't know about that PR. Let me know what you guys think and which approach is more viable.

**Comment by andrew-christianson** [comment]
@leungbk thanks for the tag. I was lagging on re-submitting emacs-lsp/lsp-python-ms#110, but it's there now as emacs-lsp/lsp-python-ms#114.  

@zeronone thanks for this, and for microsoft/pyright#810 ! This looks good, and addresses one of my TODOs on the other PR.  To confirm: are you assuming that users will set other configuration (e.g. venvPath, pythonPath) with a pyrightconfig.json?

File: lsp-pyright.el [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-stdio-connection (lambda ()
+                                          (cons (lsp-package-path 'pyright)
+                                                lsp-pyright-langserver-command-args)))
+  :major-modes '(python-mode)
+  :server-id 'pyright
+  :multi-root t
+  :priority -1
```
Comment (yyoncho) [inline lsp-pyright.el:167]: this and pyls priority should be different.

File: lsp-pyright.el [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+(defgroup lsp-pyright nil
+  "LSP support for python using the Pyright Language Server."
+  :group 'lsp-mode
+  :link '(url-link "https://github.com/microsoft/pyright"))
+
+(defcustom lsp-pyright-langserver-command-args '("--stdio")
+  "Command to start pyright-langserver."
+  :type '(repeat string)
+  :group 'lsp-pyright
+  :package-version '(lsp-mode . "7.0"))
```
Comment (yyoncho) [inline lsp-pyright.el:38]: nit: 7.0 -> 7.0.1

File: lsp-pyright.el (hunk 1 above)
Comment (zeronone) [reply lsp-pyright.el]: As it is a new lsp-client, I set priority to lower `-2`.

**Comment by zeronone** [comment]
> 2. standalone file in lsp-mode project

+1 for this option. As of now I think, it is quite simple and does not warrant a separate repo. 

----------

@andrew-christianson I added `python.pythonPath` configuration (which is basically `locate-python` function from `lsp-python-ms`). I am a little bit hesitant to add `python.venvPath` as I think it is only useful in multi-root workspaces where each project has different python version, in which case the user needs additional configuration in `pyrightconfig.json` anyway (`executionEnvironment`) - please correct me if I am wrong here!

The `locate-python` function can be improved, but for now it seems sufficient (i.e: detect `VIRTUAL_ENV` env var set by `pyvenv`)

Regarding: https://github.com/emacs-lsp/lsp-python-ms/pull/114 I actually don't how the integration would work, and how we can run two simultaneous language servers at the same time. If you think that is a better approach, I can close this PR.

**Comment by leungbk** [comment]
@seagle0128 What's the preferred option here?

**Comment by seagle0128** [comment]
> @seagle0128 What's the preferred option here?

I am fine with option 2 and 3. Two servers in `lsp-go`.

**Comment by andrew-christianson** [comment]
I'm in favor of option 2/3 here.  I like the side-by-side version in emacs-lsp/lsp-python-ms#114, but it's easy enough to make :add-on a configurable value here vs there, and it'd probably be better to wait for pyright to stabilize a bit more.  

I'll plan to close out the other PR and contribute any changes I have for pyright here.

That said, @yyoncho I still think our discussion from #1923 would be nice for this application, so we can mask out features of any language server when running as an add-on

**Comment by yyoncho** [comment]
Ok, seems like everyone is fine with 2). I will create the repo and I will send invites to @zeronone @andrew-christianson and @seagle0128 . If someone else is interested to work on that he/she could ping me.


@andrew-christianson 

> That said, @yyoncho I still think our discussion from #1923 would be nice for this application, so we can mask out features of any language server when running as an add-on

We should define what and how we want to achieve it. ATM I believe that you can alter the records in method-requirements, e. g.
changing 
``` elisp
    ("textDocument/signatureHelp" :capability :signatureHelpProvider)
```
to 
``` elisp
("textDocument/signatureHelp" 
 (lambda (workspace)
   (with-lsp-workspace workspace
     (and (lsp--capability :signatureHelpProvider)
          (not (eq (lsp--client-server-id (lsp--workspace-client workspace))
                   'mspyls))))))
```
will disable signature help for mspyls server. We can create some helper functions to make that easier but IMO, either way, the configuration should come from lsp-mode contributors and it won't be something that average user will configure.

**Comment by yyoncho** [comment]
Please move the PR here - https://github.com/emacs-lsp/lsp-pyright

**Pull request closed without merging**